jems-testing/src/main/java/org/dmfs/jems/hamcrest/matchers/StackMatcher.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.jems.hamcrest.matchers;
+
+import org.dmfs.iterables.decorators.Mapped;
+import org.dmfs.iterables.elementary.Seq;
+import org.dmfs.iterators.Function;
+import org.dmfs.jems.stack.Stack;
+import org.dmfs.optional.Optional;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+
+import static org.dmfs.jems.hamcrest.matchers.PresentMatcher.isPresent;
+import static org.hamcrest.Matchers.is;
+
+
+/**
+ * A {@link Matcher} which matches an entire {@link Stack}.
+ *
+ * @author Marten Gajda
+ */
+public final class StackMatcher<E> extends TypeSafeDiagnosingMatcher<Stack<E>>
```

Comment:
How about adapting to `Iterable` with `IterableStack`, since that is already tested in isolation, and just matching with `IterableMatcher`. Also using `FeatureMatcher` (a bit misuse, cause this is not a 'property' of `Stack`, but it should worrk anyway:
```java
    public final class StackMatcher<T> extends FeatureMatcher<Stack<T>, Iterable<T>>
    {

        StackMatcher(Stack<T> expectedStack)
        {
            super((Matcher<? super Iterable<T>>) IterableMatcher.iteratesTo(new StackIterable<T>(expectedStack)), 
                    "Elements in Stack", 
                    "Elements in Stack");
        }


        @Override
        protected Iterable<T> featureValueOf(Stack<T> actualStack)
        {
            return new StackIterable<>(actualStack);
        }
    }
```
It compiles, I haven't tried it in the test though.

Reply:
Good idea. I've updated the code with a few changes. The StackMatcherTest still passes and I can't see circular dependencies, nice.

Reply:
I had second thoughts about it afterwards, whether matching it as `StackIterable` does all the checks needed, since `Stack` is actually a more complex type. But I suppose it should be okay anyway, since `Stack` -> `Iterable` is a 'full' 'projection' here.

Reply:
That was my initial concern as well. I had the same idea yesterday. Anyway, all the tests pass so I guess it's fine.